engine/lib/phases/phase_direct_and_mut.ml

```diff
@@ -48,180 +186,199 @@ struct
-                           Some
-                             ( dty expr.span type_output0,
-                               returned_value_ident,
-                               returned_value_ident,
-                               expr.span ))
-                    @ idents)
-                    expr
-                  @@ List.fold_right
-                       ~init:
-                         (if ret_unit then UB.unit_expr expr.span
-                         else
-                           {
-                             expr with
-                             e = LocalVar returned_value_ident;
-                             typ = dty expr.span type_output0;
-                           })
-                       ~f:UB.make_seq assigns)
+                  Some (var, place_to_lhs place)
+              | _ -> None
+            in
+            let to_ty_span x =
+              let span = span_of_either x in
+              (dty span (ty_of_either x), span)
+            in
+            List.mapi ~f:(fun i -> to_ident_lhs i &&& to_ty_span) mutargs
+          in
+
+          let out_var = LocalIdent.{ id = var_id_of_int 0; name = "out" } in
+          let otype = dty f.span otype in
+          let pat =
+            let out =
+              if UB.is_unit_typ otype then []
+              else [ UB.make_var_pat out_var otype f.span ]
+            in
+            List.map
+              ~f:(function
+                | Some (var, _), (ty, span) -> UB.make_var_pat var ty span
+                | None, (ty, span) -> UB.make_wild_pat ty span)
+              mutargs
+            @ out
+            |> UB.make_tuple_pat
+          in
+          let f_call =
+            let f : B.expr =
+              let typ =
+                B.TArrow (List.map ~f:(fst >> b_ty_of_either) args, pat.typ)
+              in
+              B.{ span = f.span; typ; e = dexpr' f.span f.e }
+            in
+            let unmut_args =
+              args
+              |> List.map
+                   ~f:
+                     (fst >> function
+                      | Either.First p -> Place.to_expr p
+                      | Either.Second e -> e)
+              |> List.map ~f:dexpr
+            in
+            B.
+              {
+                e = App { f; args = unmut_args };
+                typ = pat.typ;
+                span = pat.span;
+              }
+          in
+          (* when lhs type accepts tuple, assigns will be an option instead of a list *)
+          let assigns =
+            let flatten (o, meta) = Option.map o Fn.(id &&& const meta) in
+            List.filter_map ~f:flatten mutargs
+            |> List.map ~f:(fun ((var, lhs), (typ, span)) ->
+                   let e = B.{ e = LocalVar var; span; typ } in
+                   let witness = Features.On.mutable_variable in
+                   B.{ e = Assign { lhs; e; witness }; span; typ = UB.unit_typ })
+          in
+          (* TODO: this should be greatly simplified when `lhs` type will accept tuples *)
```

Comment:
I'm stopping to add comments for them, please find all of them and make sure they are tracked in issues and the issues are referenced.